### PR TITLE
Add new entries to `TypeName`

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Current progress is: 
- Function code threaded through system but missing important pieces
- Closure is only partially threaded through
- Scalar and tuple code on [this branch](https://github.com/rust-lang/chalk/compare/master...MarkMcCaskey:feature/add-scalar-to-typename?expand=1)

Part of #368

### CODE REVIEW

File: chalk-ir/src/lib.rs [hunk 1]
```diff
@@ -96,6 +96,41 @@ impl<G: HasInterner> HasInterner for InEnvironment<G> {
...
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum FloatTy {
+    F32,
+    F64,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Scalar {
```
Comment (jackh726) [inline chalk-ir/src/lib.rs:126]: I like this idea of grouping these into a single enum. 👍

File: chalk-ir/src/lib.rs [hunk 2]
```diff
@@ -146,6 +193,12 @@ impl UniverseIndex {
 #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct StructId<I: Interner>(pub I::DefId);
 
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct FnDefId<I: Interner>(pub I::DefId);
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ClosureId<I: Interner>(pub I::DefId);
```
Comment (jackh726) [inline chalk-ir/src/lib.rs:200]: We had previously talked about moving all of these as associated types on `Interner`. This is making me somewhat think this is a good idea. But maybe not in this PR.

File: chalk-parse/src/ast.rs [hunk 3]
```diff
@@ -41,6 +42,34 @@ pub struct StructFlags {
     pub fundamental: bool,
 }
 
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct FnDefn {
+    pub name: Identifier,
+    pub parameter_kinds: Vec<ParameterKind>,
+    pub where_clauses: Vec<QuantifiedWhereClause>,
+    pub params: Vec<Field>,
+    pub returns: Vec<Field>,
```
Comment (jackh726) [inline chalk-parse/src/ast.rs:51]: This doesn't feel right to me. If anything, it should be a single `Field`.

File: chalk-parse/src/ast.rs [hunk 4]
```diff
@@ -41,6 +42,34 @@ pub struct StructFlags {
...
+pub struct FnFlags {
+    pub upstream: bool,
+    pub fundamental: bool,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct ClosureDefn {
+    // What's a closure's name? Need to generate an `Identifier` or something
+    // TODO(markmccaskey): investigate what rustc does here
+    pub name: Identifier,
```
Comment (jackh726) [inline chalk-parse/src/ast.rs:65]: Technically, closures aren't nameable. But I'm not sure if rustc generates a name.

File: chalk-parse/src/ast.rs [hunk 5]
```diff
@@ -41,6 +42,34 @@ pub struct StructFlags {
     pub fundamental: bool,
 }
 
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct FnDefn {
+    pub name: Identifier,
+    pub parameter_kinds: Vec<ParameterKind>,
```
Comment (jackh726) [inline chalk-parse/src/ast.rs:48]: Might be better to rename this "generics" or something.

File: chalk-rust-ir/src/lib.rs [hunk 6]
```diff
@@ -90,6 +90,44 @@ pub struct StructFlags {
...
+    //pub binders: Binders<StructDatumBound<I>>,
+    pub id: ClosureId<I>,
+}
+
+impl<I: Interner> ClosureDatum<I> {
+    pub fn name(&self, interner: &I) -> TypeName<I> {
+        self.id.cast(interner)
+    }
+}
+
```
Comment (jackh726) [inline chalk-rust-ir/src/lib.rs:130]: Why are these here, but the others are in `chalk-ir`? Plus there's duplicate structs like `FnDefFlags`/`FnFlags`

File: chalk-solve/src/clauses/program_clauses.rs [hunk 7]
```diff
@@ -275,6 +275,172 @@ impl<I: Interner> ToProgramClauses<I> for StructDatum<I> {
     }
 }
 
+impl<I: Interner> ToProgramClauses<I> for FnDefDatum<I> {
```
Comment (jackh726) [inline chalk-solve/src/clauses/program_clauses.rs:278]: This seems mostly duplicated from `StructDatum`?

File: chalk-parse/src/ast.rs (hunk 3 above)
Comment (flodiebold) [reply chalk-parse/src/ast.rs:51]: Do you actually need the function params and return at all? I don't see where they should matter to Chalk :thinking:

Comment (nikomatsakis) [reply chalk-parse/src/ast.rs:51]: Well, we will *eventually* need this information. We are going to have to generate `FnMut` impls and the like.

File: chalk-parse/src/ast.rs (hunk 4 above)
Comment (nikomatsakis) [reply chalk-parse/src/ast.rs:65]: Yes, rustc has an internal name. I think giving them a dummy name is a fine thing to do for chalk tests.

File: chalk-parse/src/ast.rs (hunk 3 above)
Comment (nikomatsakis) [reply chalk-parse/src/ast.rs:51]: (But it might plausibly be done in a follow-up PR)

File: chalk-parse/src/ast.rs (hunk 4 above)
Comment (nikomatsakis) [reply chalk-parse/src/ast.rs:65]: That said, we'll probably want to (a) consolidate some of this with functions and (b) we may want other things eventually to better match rustc. But we can evolve this part.

File: chalk-solve/src/clauses/program_clauses.rs (hunk 7 above)
Comment (nikomatsakis) [reply chalk-solve/src/clauses/program_clauses.rs:278]: Interesting stuff! We'll have to think about just which rules we want here...

File: chalk-rust-ir/src/lib.rs (hunk 6 above)
Comment (MarkMcCaskey) [reply chalk-rust-ir/src/lib.rs:130]: The current implementation is pretty much just a duplication and slight modification of `Struct` as a starting point.  For example, [`chalk_ast::StructFlags`](http://rust-lang.github.io/chalk/chalk_parse/ast/struct.StructFlags.html) and [`chalk_rust_ir::StructFlags`](http://rust-lang.github.io/chalk/chalk_rust_ir/struct.StructFlags.html) turned into `FnDefFlags` and `FnFlags`.  And `FnDefDatum` is here because [`StructDatum`](http://rust-lang.github.io/chalk/chalk_rust_ir/struct.StructDatum.html) is also here!

I'm oscillating between thinking I understand the goal and the pieces I'm working with and thinking I don't understand it :smile: .  I'll go back to the basics and see if I can get a better understanding of the pieces involved here!  The review helped me realize that my working understanding of [`Field`](http://rust-lang.github.io/chalk/chalk_parse/ast/struct.Field.html) was off base, for example, and perhaps I've missed other important details!

**Comment by MarkMcCaskey** [comment]
Thanks for the review!  I'm realizing that I've probably gone a bit too far into unknown territory, so I'm going to read through the docs on chalk and think about this some more to try to get a more solid foundation before continuing tomorrow night.

**Comment by nikomatsakis** [comment]
@MarkMcCaskey perhaps, but actually I thnk the code you wrote is all basically "on target" -- that is, the details of closures/functions might be a bit off, but you've got basically the right pieces in basically the right places :)

**Comment by MarkMcCaskey** [comment]
I added basic parsing for (empty) function definitions so we can have tests for this.

I noticed in the tests there's code like,

```rust
            impl<T> Print for OnlySized<T> {
                // fn print() {
                //     println!("{}", std::mem::size_of::<T>());
                // }
            }
``` 

Is it in the scope of this PR to make this work?

**Comment by nikomatsakis** [comment]
@MarkMcCaskey definitely not =)

**Comment by nikomatsakis** [comment]
Closing in favor of #394 -- let's revisit this stuff once that one lands.
